### PR TITLE
Refactor the code related to loss computation

### LINKED ISSUE #854: [BUG] Loss signatures: CE Loss failure because of additional `params` argument
### Describe the bug
During [loss computations](https://github.com/mlcommons/GaNDLF/blob/master/GANDLF/compute/loss_and_metric.py#L113)
 it is assumed that loss function takes three params: prediction, target, params. However, it's not true for [CE](https://github.com/mlcommons/GaNDLF/blob/master/GANDLF/losses/regression.py#L59) loss that takes only prediction and target, so using `loss_function: ce` fails.
 
### To Reproduce
try to train any model with `loss_function: ce`

### Media
If applicable, add images, screenshots or other relevant media to help explain your problem.

### Environment information

GaNDLF version, OS, and any other relevant information.

### Additional context
The straightforward solution is just to add an unused `params` arg to CE function.  However, I believe, doing this would cause linter / codacy failures as parameter is defined but not used. In this case the best option IMO is to create a standard class interface for losses:

```python
## different file, so that this can be used by segmentation, classification, regression, and compound/hybrid losses

class LossFunctionInterface(ABC):
    @staticmethod
    @abstractmethod
    def compute_loss(
        self, targets: torch.Tensor, predictions: torch.Tensor, params: dict
    ):
        """
        Abstract method to compute the loss function.

        Args:
            targets (torch.Tensor): The target values.
            predictions (torch.Tensor): The predicted values.
            params (dict): The parameters.

        Raises:
            NotImplementedError: This method must be implemented in the derived class.
        """
        raise NotImplementedError

class DCCE(LossInterface):
    @staticmethod
    def compute_loss(predictions: torch.Tensor, targets: torch.Tensor, params: dict) -> torch.Tensor:
        ### ... move DCCE calculation logic there...

# in loss_and_metric.py it can be used as:
# loss_function.calc(predictions, targets, params)
```
and the same with all other losses. In this case all the losses would have the same signature and can be used interchangeably. If signature of any loss function differs, both Codacy and IDE would warn developer that something goes wrong.

### PR DESCRIPTION
**Pull request by szmazurek**
Fixes #854

## Proposed Changes
- create a common interface for the loss functions
- refactor the codebase to use new loss implementations

## Checklist

- [ ] [`CONTRIBUTING`](https://github.com/mlcommons/GaNDLF/blob/master/CONTRIBUTING.md) guide has been followed.
- [ ] PR is based on the [current GaNDLF master ](https://docs.github.com/en/desktop/contributing-and-collaborating-using-github-desktop/keeping-your-local-repository-in-sync-with-github/syncing-your-branch-in-github-desktop?platform=windows).
- [ ] Non-breaking change (does **not** break existing functionality): provide **as many** details as possible for _any_ breaking change.
- [ ] Function/class source code documentation added/updated (ensure `typing` is used to provide type hints, including and not limited to using `Optional` if a variable has a pre-defined value).
- [ ] Code has been [blacked](https://github.com/psf/black#usage) for style consistency and linting.
- [ ] If applicable, version information [has been updated in GANDLF/version.py](https://github.com/mlcommons/GaNDLF/blob/master/GANDLF/version.py).
- [ ] If adding a git submodule, add to list of exceptions for black styling in [pyproject.toml](https://github.com/mlcommons/GaNDLF/blob/master/pyproject.toml) file.
- [ ] [Usage documentation](https://github.com/mlcommons/GaNDLF/blob/master/docs) has been updated, if appropriate.
- [ ] Tests added or modified to [cover the changes](https://app.codecov.io/gh/mlcommons/GaNDLF); if coverage is reduced, please give explanation.
- [ ] If customized dependency installation is required (i.e., a separate `pip install` step is needed for PR to be functional), please ensure it is reflected in all the files that control the CI, namely: [python-test.yml](https://github.com/mlcommons/GaNDLF/blob/master/.github/workflows/python-test.yml), and all docker files [[1](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-CPU),[2](https://github.com/mlcommons/GaNDLF/blob/devcontainer_build_fix/Dockerfile-CUDA11.6),[3](https://github.com/mlcommons/GaNDLF/blob/master/Dockerfile-ROCm)].
- [ ] The `logging` library is being used and no `print` statements are left.

### CODE REVIEW

**Comment by szmazurek** [comment]
@sarthakpati What is your opinion on the implementation of interface and example loss? When we agree on that I will proceed to port the rest.

**Comment by sarthakpati** [comment]
Should this PR not be going into https://github.com/szmazurek/GaNDLF/tree/dev ?

**Comment by szmazurek** [comment]
> Should this PR not be going into https://github.com/szmazurek/GaNDLF/tree/dev ?

Could be, but as it is directly related to existing core code I tought that we can update directly here.

**Comment by sarthakpati** [comment]
[This commit](https://github.com/mlcommons/GaNDLF/pull/965/commits/ec3670dad47f683c86bdf72db0c7932bbddc1929) was done without proper git initialization and hence is getting flagged by the CLA bot. Can you rebase this?

**Comment by szmazurek** [comment]
> [This commit](https://github.com/mlcommons/GaNDLF/pull/965/commits/ec3670dad47f683c86bdf72db0c7932bbddc1929) was done without proper git initialization and hence is getting flagged by the CLA bot. Can you rebase this?

Done

File: GANDLF/losses/loss_interface.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        for class_idx in range(self.num_classes):
+            current_loss = self._compute_single_class_loss(
+                prediction, target, class_idx
+            )
+            accumulated_loss += (
+                self._optional_loss_operations(current_loss)
+                * self.penalty_weights[class_idx]
+            )
+
+        # TODO shouldn't we always divide by the number of classes?
```
[the remark this answers is not in the document]
Comment (szmazurek) [reply GANDLF/losses/loss_interface.py:69]: Yeah we can actually go everytime with division, as for single class the division will simply be by 1, so no changes

File: GANDLF/losses/loss_interface.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    ) -> torch.Tensor:
+        """Compute loss for a pair of prediction and target tensors. To be implemented by child classes."""
+        pass
+
+    def forward(self, prediction: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        accumulated_loss = torch.tensor(0.0, device=prediction.device)
+
+        for class_idx in range(self.num_classes):
+            current_loss = self._compute_single_class_loss(
+                prediction, target, class_idx
```
[the remark this answers is not in the document]
Comment (szmazurek) [reply GANDLF/losses/loss_interface.py:62]: I wanted to keep consistent with previous loss implementations, so I am performing loss computation separately for each class and later accumulate.

File: GANDLF/losses/loss_interface.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sarthakpati) [reply GANDLF/losses/loss_interface.py:69]: Sounds good!

File: GANDLF/losses/loss_interface.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sarthakpati) [reply GANDLF/losses/loss_interface.py:62]: Got it, thanks for the clarification.

**Review by sarthakpati — APPROVED** [review]

**Pull request merged**